src/tamp_improv/approaches/mpc.py

```diff
@@ -0,0 +1,221 @@
+                (
+                    control_points[:, dim]
+                    if len(control_points.shape) > 1
+                    else control_points
+                ),
+            )
+
+        return np.clip(trajectory, box_space.low, box_space.high)
+
+    def _sample_from_nominal(self, nominal: np.ndarray) -> list[np.ndarray]:
+        """Sample new trajectories around nominal."""
+        assert self._env is not None
+
+        if self._is_discrete:
+            # For discrete actions, flip with probability noise_scale
+            trajectories = []
+            for _ in range(self.config.num_rollouts - 1):
+                flip_mask = (
+                    self._rng.random(size=self.config.horizon) < self.config.noise_scale
+                )
+                new_traj = nominal.copy()
+                new_traj[flip_mask] = 1 - new_traj[flip_mask]  # Flip 0->1 or 1->0
+                trajectories.append(new_traj)
+            return trajectories
+
+        if not self._box_space:
+            raise ValueError("Unsupported action space type")
+
+        # For continuous actions, use spline interpolation with noise
+        box_space = cast(gym.spaces.Box, self._env.action_space)
+        nominal_control_points = np.array(
+            [nominal[int(t)] for t in self._control_times]
+        )
+
+        action_shape = box_space.shape if box_space.shape else ()
+        noise = self._rng.normal(
+            loc=0,
+            scale=self.config.noise_scale,
+            size=(self.config.num_rollouts - 1, self.config.num_control_points)
+            + action_shape,
+        )
+        new_control_points = nominal_control_points + noise
+
+        trajectories = []
+        for points in new_control_points:
+            trajectory = np.zeros((self.config.horizon,) + action_shape)
+            for dim in range(points.shape[-1] if len(points.shape) > 1 else 1):
+                idx = (..., dim) if len(points.shape) > 1 else ...
+                trajectory[idx] = np.interp(
+                    self._trajectory_times, self._control_times, points[idx]
+                )
+            trajectory = np.clip(trajectory, box_space.low, box_space.high)
+            trajectories.append(trajectory)
+        return trajectories
+
+    def _score_trajectory(
+        self, trajectory: np.ndarray, init_obs: ObsType_contra
+    ) -> float:
+        """Score a trajectory by simulating it."""
+        assert self._env is not None
+
+        obs = init_obs
+        total_reward = 0.0
+
+        self._env.reset(options={"initial_obs": obs})
+
+        for action in trajectory:
+            if self._is_discrete:
+                action = int(action)
+            obs, reward, terminated, truncated, _ = self._env.step(action)
+            total_reward += float(reward)
+            if terminated or truncated:
+                break
+
+        return total_reward
+
+    def save(self, path: str) -> None:
+        """Save policy parameters."""
```

Comment:
hmm interesting. I actually think we should save and load nothing here. MPC doesn't learn anything. and we typically don't assume that saving and loading happens in the same state. so the last solution that we saved might be garbage in the state where we load. and the config stuff similarly doesn't change as the policy goes on, so there's no need to save it.